lib/happo/snapshot_comparer.rb

```diff
@@ -27,65 +32,89 @@ def compare!
-      # more work. This might happen if some of the headers are different.
-      return no_diff if array_before == array_after
+      max_height = [@png_after.height, @png_before.height].max
+      max_width = [@png_after.width, @png_before.width].max
 
-      sdiff = Diff::LCS.sdiff(array_before, array_after)
-      number_of_different_rows = 0
+      diff_image = ChunkyPNG::Image.new(max_width, max_height)
 
-      sprite, all_comparisons = initialize_comparison_images(
-        [@png_after.width, @png_before.width].max, sdiff.size)
+      number_of_different_pixels = 0
 
-      sdiff.each_with_index do |row, y|
-        # each row is a Diff::LCS::ContextChange instance
-        all_comparisons.each { |image| image.render_row(y, row) }
-        number_of_different_rows += 1 unless row.unchanged?
+      max_height.times do |y|
+        max_width.times do |x|
+          pixel_before = @png_before.get_pixel(x, y)
+          pixel_after = @png_after.get_pixel(x, y)
+          number_of_different_pixels += 1 unless pixel_before == pixel_after
+          diff_image.set_pixel(x, y, faded_magenta_pixel(
+            pixel_after  || TRANSPARENT,
+            pixel_before || TRANSPARENT
+          ))
+        end
       end
 
-      percent_changed = number_of_different_rows.to_f / sdiff.size * 100
+      # Even though we have a check at the top to see image equality, there's
+      # still a chance we end up here with no difference. This might happen if
+      # some of the headers are different.
+      return no_diff if number_of_different_pixels.zero?
+
+      percent_changed =
+        number_of_different_pixels.to_f / (max_width * max_height * 100)
+
       {
         diff_in_percent: percent_changed,
-        diff_image:      (sprite if percent_changed > 0),
+        diff_image:      (diff_image if percent_changed.positive?),
       }
     end
 
     private
 
-    # @param [ChunkyPNG::Image]
-    # @return [Array<Array<Integer>>]
-    def to_array_of_arrays(chunky_png)
-      array_of_arrays = []
-      chunky_png.height.times do |y|
-        array_of_arrays << chunky_png.row(y)
+    # Compute a score that represents the difference between 2 pixels
+    #
+    # This method simply takes the Euclidean distance between the RGBA channels
+    # of 2 colors over the maximum possible Euclidean distance. This gives us a
+    # percentage of how different the two colors are.
+    #
+    # Although it would be more perceptually accurate to calculate a proper
+    # Delta E in Lab colorspace, we probably don't need perceptual accuracy for
+    # this application, and it is nice to avoid the overhead of converting RGBA
+    # to Lab.
+    #
+    # @param pixel_after [Integer]
+    # @param pixel_before [Integer]
+    # @return [Float] number between 0 and 1 where 1 is completely different
+    #   and 0 is no difference
+    def pixel_diff_score(pixel_after, pixel_before)
+      ChunkyPNG::Color::euclidean_distance_rgba(pixel_after, pixel_before) /
```

Comment:
Yeah, I think it would improve perf a little. But I think we should just add it to chunkypng instead. https://github.com/wvanbergen/chunky_png/blob/cf9cf86894c4490421bb73263a8379ae4f2abfab/lib/chunky_png/color.rb#L716 I'll submit a PR over there.